driver-sync/src/test/functional/com/mongodb/client/RetryableWritesProseTest.java

```diff
@@ -0,0 +1,210 @@
+
+// See https://github.com/mongodb/specifications/tree/master/source/retryable-writes/tests/README.rst#replica-set-failover-test
+public class RetryableWritesProseTest extends DatabaseTestCase {
+    private MongoClient clientUnderTest;
+    private MongoClient failPointClient;
+    private MongoClient stepDownClient;
+    private final TestCommandListener commandListener;
+    private final String databaseName;
+    private final String collectionName;
+    private final String description;
+
+    public RetryableWritesProseTest() {
+        this.description = "Replica set failover test for retryable writes";
+        this.databaseName = getDefaultDatabaseName();
+        this.collectionName = "test";
+        this.commandListener = new TestCommandListener();
+    }
+
+    @BeforeClass
+    public static void beforeClass() {
+    }
+
+    @AfterClass
+    public static void afterClass() {
+    }
+
+    @Before
+    @Override
+    public void setUp() {
+        assumeTrue(canRunTests());
+
+        setUpClientUnderTest();
+        setUpFailPointClient();
+        setUpStepDownClient();
+    }
+
+    @After
+    public void cleanUp() {
+        if (failPointClient != null) {
+            MongoDatabase failPointAdminDB = failPointClient.getDatabase("admin");
+            failPointAdminDB.runCommand(
+                    new Document("configureFailPoint", "stepdownHangBeforePerformingPostMemberStateUpdateActions")
+                    .append("mode", "off"));
+        }
+
+        if (clientUnderTest != null) {
+            clientUnderTest.close();
+        }
+        if (stepDownClient != null) {
+            stepDownClient.close();
+        }
+    }
+
+    @Test
+    public void shouldPassAllOutcomes() {
+        // Using the client under test, insert a document and observe a successful write result.
+        // This will ensure that initial discovery takes place.
+        MongoCollection<Document> collection = insertInitialDocument();
+
+        // Using the fail point client, activate the fail point by setting mode to "alwaysOn".
+        activateFailPoint();
+
+        // Using the step down client, step down the primary by executing the command { replSetStepDown: 60, force: true}.
+        // This operation will hang so long as the fail point is activated. When the fail point is later deactivated,
+        // the step down will complete and the primary's client connections will be dropped. At that point, any ensuing
+        // network error should be ignored.
+        MongoDatabase stepDownDB = stepDownPrimary();
+
+        // Using the client under test, insert a document and observe a successful write result. The test MUST assert
+        // that the insert command fails once against the stepped down node and is successfully retried on the newly
+        // elected primary (after SDAM discovers the topology change). The test MAY use APM or another means to observe
+        // both attempts.
+        insertDocument(collection);
```

Comment:
You could check the commandListeners' events to see if one was an error.
There should be no errors before the insert and one after it.

Reply:
Done.